File: desktop/src/api/rutas.ts
import { invoke } from "@tauri-apps/api/core";
import { solicitarSincronizacionNube } from "../eventosNube";

// Espejo de comandos/rutas.rs y dto/rutas.rs — ver también
// src/application/rutas.rs y src/models/{vehiculo_ruta,encargado_ruta,salida_ruta}.rs
// del núcleo. Dominio de "Control de rutas"
// (docs/planes-implementados/plan-control-rutas.md).

export type ResultadoSalidaRuta = "Permitido" | "PermitidoConAutorizacion";

export interface VehiculoRuta {
  id: number;
  numero_unidad: string | null;
  placa: string;
  activo: boolean;
}

export interface EncargadoRuta {
  id: number;
  codigo_empleado: string;
  nombre: string;
  /** Nunca se captura del lado de escritorio -- ver dto/rutas.rs. */
  cedula: string | null;
  activo: boolean;
}

export interface DatosVehiculoRuta {
  numero_unidad: string | null;
  placa: string;
  activo: boolean;
}

export interface DatosEncargadoRuta {
  codigo_empleado: string;
  nombre: string;
  activo: boolean;
}

/** Catálogo de números de ruta válidos -- a diferencia de
 * vehículos/encargados (consultivos), éste SÍ restringe: `registrarSalidaRuta`
 * rechaza cualquier número que no exista o esté dado de baja acá. */
export interface Ruta {
  id: number;
  numero: number;
  activo: boolean;
}

/** Para la grilla de administración -- trae activos e inactivos.
 * `listarVehiculosRutaSeleccionables` es la contraparte para el selector de
 * salida de ruta, donde un vehículo desactivado nunca es una opción válida
 * -- esa decisión la toma el núcleo, no quien llama. */
export function listarVehiculosRuta(): Promise<VehiculoRuta[]> {
  return invoke("listar_vehiculos_ruta");
}

export function listarVehiculosRutaSeleccionables(): Promise<VehiculoRuta[]> {
  return invoke("listar_vehiculos_ruta_seleccionables");
}

export async function crearVehiculoRuta(datos: DatosVehiculoRuta): Promise<number> {
  const id = await invoke<number>("crear_vehiculo_ruta", { datos });
  solicitarSincronizacionNube();
  return id;
}

export async function actualizarVehiculoRuta(id: number, datos: DatosVehiculoRuta): Promise<void> {
  await invoke("actualizar_vehiculo_ruta", { id, datos });
  solicitarSincronizacionNube();
}

/** Para la grilla de administración -- trae activos e inactivos.
 * `listarEncargadosRutaSeleccionables` es la contraparte para un selector
 * de wizard (elegir encargado para una salida nueva), donde uno desactivado
 * nunca es una opción válida -- esa decisión la toma el núcleo. */
export function listarEncargadosRuta(): Promise<EncargadoRuta[]> {
  return invoke("listar_encargados_ruta");
}

export function listarEncargadosRutaSeleccionables(): Promise<EncargadoRuta[]> {
  return invoke("listar_encargados_ruta_seleccionables");
}

export async function crearEncargadoRuta(datos: DatosEncargadoRuta): Promise<number> {
  const id = await invoke<number>("crear_encargado_ruta", { datos });
  solicitarSincronizacionNube();
  return id;
}

export async function actualizarEncargadoRuta(
  id: number,
  datos: DatosEncargadoRuta,
): Promise<void> {
  await invoke("actualizar_encargado_ruta", { id, datos });
  solicitarSincronizacionNube();
}

/** Para la grilla de administración -- trae activas y dadas de baja.
 * `listarRutasSeleccionables` es la contraparte para el selector de salida
 * de ruta, donde una ruta dada de baja nunca es una opción válida -- esa
 * decisión la toma el núcleo. */
export function listarRutas(): Promise<Ruta[]> {
  return invoke("listar_rutas");
}

export function listarRutasSeleccionables(): Promise<Ruta[]> {
  return invoke("listar_rutas_seleccionables");
}

export async function crearRuta(numero: number): Promise<number> {
  const id = await invoke<number>("crear_ruta", { numero });
  solicitarSincronizacionNube();
  return id;
}

export async function crearRutasRango(desde: number, hasta: number): Promise<number[]> {
  const ids = await invoke<number[]>("crear_rutas_rango", { desde, hasta });
  solicitarSincronizacionNube();
  return ids;
}

export async function darDeBajaRuta(id: number): Promise<void> {
  await invoke("dar_de_baja_ruta", { id });
  solicitarSincronizacionNube();
}

export async function reactivarRuta(id: number): Promise<void> {
  await invoke("reactivar_ruta", { id });
  solicitarSincronizacionNube();
}

/** Fila para la pantalla "Rutas activas" -- análoga a
 * `MovimientoVisitaActivoResumen`. */
export interface SalidaRutaActivaResumen {
  id: number;
  vehiculo_placa: string;
  vehiculo_numero_unidad: string | null;
  encargado_nombre: string;
  numero_ruta: number;
  sub_numero: number;
  numero_documento: string;
  /** `YYYY-MM-DD`. */
  fecha_documento: string;
  resultado: ResultadoSalidaRuta;
  /** ISO 8601 (UTC) — convertir con `new Date(...)` antes de mostrar. */
  fecha_hora_salida: string;
  usuario_salida_nombre: string;
}

/** Espejo de `SolicitudSalidaRutaEntrada` (dto/rutas.rs) --
 * `usuario_salida_id`/`fecha_hora_salida` no viajan: el núcleo los fija con
 * el actor y el reloj validados de la transacción, nunca con lo que mande
 * el webview. */
export interface SolicitudSalidaRuta {
  vehiculo_placa: string;
  vehiculo_numero_unidad: string | null;
  encargado_nombre: string;
  encargado_codigo_empleado: string | null;
  numero_ruta: number;
  sub_numero: number;
  numero_documento: string;
  /** `YYYY-MM-DD`. */
  fecha_documento: string;
  tiene_correo_autorizacion: boolean;
}

export interface RetornoSalidaRuta {
  /** ISO 8601 (UTC). */
  fecha_hora: string;
  usuario_id: number;
}

/** Registro completo -- lo usa la pantalla de detalle/confirmación de
 * retorno, a diferencia de `SalidaRutaActivaResumen` (fila de la grilla). */
export interface SalidaRuta {
  id: number;
  vehiculo_id: number | null;
  vehiculo_placa: string;
  vehiculo_numero_unidad: string | null;
  encargado_id: number | null;
  encargado_nombre: string;
  numero_ruta: number;
  sub_numero: number;
  numero_documento: string;
  /** `YYYY-MM-DD`. */
  fecha_documento: string;
  resultado: ResultadoSalidaRuta;
  /** ISO 8601 (UTC). */
  fecha_hora_salida: string;
  usuario_salida_id: number;
  retorno: RetornoSalidaRuta | null;
}

export async function registrarSalidaRuta(solicitud: SolicitudSalidaRuta): Promise<number> {
  const id = await invoke<number>("registrar_salida_ruta", { solicitud });
  solicitarSincronizacionNube();
  return id;
}

export async function registrarRetornoRuta(salidaId: number): Promise<void> {
  await invoke("registrar_retorno_ruta", { salidaId });
  solicitarSincronizacionNube();
}

export function listarRutasActivas(): Promise<SalidaRutaActivaResumen[]> {
  return invoke("listar_rutas_activas");
}

export function buscarSalidaRuta(id: number): Promise<SalidaRuta | null> {
  return invoke("buscar_salida_ruta", { id });
}

File: desktop/src-tauri/src/comandos/rutas.rs
use control_acceso::mensajes::{
    mensaje_encargado_ruta, mensaje_ruta, mensaje_ruta_catalogo, mensaje_vehiculo_ruta,
};
use control_acceso::models::encargado_ruta::EncargadoRuta;
use control_acceso::models::ruta::Ruta;
use control_acceso::models::salida_ruta::{SalidaRuta, SalidaRutaActivaResumen};
use control_acceso::models::vehiculo_ruta::VehiculoRuta;

use crate::dto::rutas::{
    DatosEncargadoRutaEntrada, DatosVehiculoRutaEntrada, SolicitudSalidaRutaEntrada,
};
use crate::estado::GuiState;

// ---- Catálogo: vehículos ----

#[tauri::command]
pub fn listar_vehiculos_ruta(state: tauri::State<GuiState>) -> Result<Vec<VehiculoRuta>, String> {
    state.sesion_activa()?;
    // Sin `mensaje_*` propio a propósito, mismo criterio que
    // `listar_visitas_activas`: es una lectura simple que devuelve
    // `DatabaseError` directo, no un `*ServiceError` de negocio.
    state
        .core()
        .listar_vehiculos_ruta()
        .map_err(|_| "No se pudo cargar la lista de vehículos".to_string())
}

/// Para el selector de salida de ruta -- un vehículo desactivado nunca es
/// una opción válida, ver `VehiculoRutaService::listar_seleccionables`.
#[tauri::command]
pub fn listar_vehiculos_ruta_seleccionables(
    state: tauri::State<GuiState>,
) -> Result<Vec<VehiculoRuta>, String> {
    state.sesion_activa()?;
    state
        .core()
        .listar_vehiculos_ruta_seleccionables()
        .map_err(|_| "No se pudo cargar la lista de vehículos".to_string())
}

#[tauri::command]
pub fn crear_vehiculo_ruta(
    datos: DatosVehiculoRutaEntrada,
    state: tauri::State<GuiState>,
) -> Result<i64, String> {
    let sesion = state.sesion_activa()?;
    state
        .core()
        .crear_vehiculo_ruta(&sesion, &datos.construir(0))
        .map_err(mensaje_vehiculo_ruta)
}

#[tauri::command]
pub fn actualizar_vehiculo_ruta(
    id: i64,
    datos: DatosVehiculoRutaEntrada,
    state: tauri::State<GuiState>,
) -> Result<(), String> {
    let sesion = state.sesion_activa()?;
    state
        .core()
        .actualizar_vehiculo_ruta(&sesion, &datos.construir(id))
        .map_err(mensaje_vehiculo_ruta)
}

// ---- Catálogo: encargados (personal KOF) ----

#[tauri::command]
pub fn listar_encargados_ruta(state: tauri::State<GuiState>) -> Result<Vec<EncargadoRuta>, String> {
    state.sesion_activa()?;
    state
        .core()
        .listar_encargados_ruta()
        .map_err(|_| "No se pudo cargar la lista de encargados".to_string())
}

/// Para un selector de wizard (entrega de gafete provisional, salida de
/// ruta) -- un encargado desactivado nunca es una opción válida, ver
/// `EncargadoRutaService::listar_seleccionables`.
#[tauri::command]
pub fn listar_encargados_ruta_seleccionables(
    state: tauri::State<GuiState>,
) -> Result<Vec<EncargadoRuta>, String> {
    state.sesion_activa()?;
    state
        .core()
        .listar_encargados_ruta_seleccionables()
        .map_err(|_| "No se pudo cargar la lista de encargados".to_string())
}

#[tauri::command]
pub fn crear_encargado_ruta(
    datos: DatosEncargadoRutaEntrada,
    state: tauri::State<GuiState>,
) -> Result<i64, String> {
    let sesion = state.sesion_activa()?;
    state
        .core()
        .crear_encargado_ruta(&sesion, &datos.construir(0))
        .map_err(mensaje_encargado_ruta)
}

#[tauri::command]
pub fn actualizar_encargado_ruta(
    id: i64,
    datos: DatosEncargadoRutaEntrada,
    state: tauri::State<GuiState>,
) -> Result<(), String> {
    let sesion = state.sesion_activa()?;
    state
        .core()
        .actualizar_encargado_ruta(&sesion, &datos.construir(id))
        .map_err(mensaje_encargado_ruta)
}

// ---- Catálogo: números de ruta ----

#[tauri::command]
pub fn listar_rutas(state: tauri::State<GuiState>) -> Result<Vec<Ruta>, String> {
    state.sesion_activa()?;
    state
        .core()
        .listar_rutas()
        .map_err(|_| "No se pudo cargar la lista de rutas".to_string())
}

/// Para el selector de salida de ruta -- una ruta dada de baja nunca es una
/// opción válida, ver `RutaCatalogoService::listar_seleccionables`.
#[tauri::command]
pub fn listar_rutas_seleccionables(state: tauri::State<GuiState>) -> Result<Vec<Ruta>, String> {
    state.sesion_activa()?;
    state
        .core()
        .listar_rutas_seleccionables()
        .map_err(|_| "No se pudo cargar la lista de rutas".to_string())
}

#[tauri::command]
pub fn crear_ruta(numero: i64, state: tauri::State<GuiState>) -> Result<i64, String> {
    let sesion = state.sesion_activa()?;
    state
        .core()
        .crear_ruta(&sesion, numero)
        .map_err(mensaje_ruta_catalogo)
}

#[tauri::command]
pub fn crear_rutas_rango(
    desde: i64,
    hasta: i64,
    state: tauri::State<GuiState>,
) -> Result<Vec<i64>, String> {
    let sesion = state.sesion_activa()?;
    state
        .core()
        .crear_rutas_rango(&sesion, desde, hasta)
        .map_err(mensaje_ruta_catalogo)
}

#[tauri::command]
pub fn dar_de_baja_ruta(id: i64, state: tauri::State<GuiState>) -> Result<(), String> {
    let sesion = state.sesion_activa()?;
    state
        .core()
        .dar_de_baja_ruta(&sesion, id)
        .map_err(mensaje_ruta_catalogo)
}

#[tauri::command]
pub fn reactivar_ruta(id: i64, state: tauri::State<GuiState>) -> Result<(), String> {
    let sesion = state.sesion_activa()?;
    state
        .core()
        .reactivar_ruta(&sesion, id)
        .map_err(mensaje_ruta_catalogo)
}

// ---- Operación: salida / retorno ----

#[tauri::command]
pub fn registrar_salida_ruta(
    solicitud: SolicitudSalidaRutaEntrada,
    state: tauri::State<GuiState>,
) -> Result<i64, String> {
    let sesion = state.sesion_activa()?;
    state
        .core()
        .registrar_salida_ruta(&sesion, solicitud.construir())
        .map(|resultado| resultado.salida_id)
        .map_err(mensaje_ruta)
}

#[tauri::command]
pub fn registrar_retorno_ruta(salida_id: i64, state: tauri::State<GuiState>) -> Result<(), String> {
    let sesion = state.sesion_activa()?;
    state
        .core()
        .registrar_retorno_ruta(&sesion, salida_id)
        .map_err(mensaje_ruta)
}

#[tauri::command]
pub fn listar_rutas_activas(
    state: tauri::State<GuiState>,
) -> Result<Vec<SalidaRutaActivaResumen>, String> {
    state.sesion_activa()?;
    state.core().listar_rutas_activas().map_err(mensaje_ruta)
}

#[tauri::command]
pub fn buscar_salida_ruta(
    id: i64,
    state: tauri::State<GuiState>,
) -> Result<Option<SalidaRuta>, String> {
    state.sesion_activa()?;
    state.core().buscar_salida_ruta(id).map_err(mensaje_ruta)
}
